### PR TITLE
[CDAP-8418] Tags management in app/stream/dataset Overview/Detailed view

### PR DESCRIPTION
**Pull request by tonybach**
JIRA: https://issues.cask.co/browse/CDAP-8418

### CODE REVIEW

File: cdap-ui/app/cdap/components/Tags/index.js [hunk 1]
```diff
@@ -15,33 +15,123 @@
...
+
+    MyMetadataApi.getTags(systemParams)
+      .combineLatest(MyMetadataApi.getTags(userParams))
+      .subscribe((res) => {
+        this.setState({
+          systemTags: res[0],
+          userTags: res[1]
+        });
+      }, (err) => {
+        console.log('Error', err);
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/Tags/index.js]: We should probably surface this error. `Unable to fetch tags` or something similar to that .

File: cdap-ui/app/cdap/components/Tags/index.js [hunk 2]
```diff
@@ -15,33 +15,123 @@
...
+      });
   }
-  componentWillReceiveProps(nextProps) {
-    let {entity} = nextProps;
-    if (entity.id !== this.props.entity.id) {
+
+  componentWillUnmount() {
+    Mousetrap.unbind('enter');
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/Tags/index.js]: We should probably bind ESC key to close the tags textbox if no value is entered in it. We used to have that functionality in the old UI.

File: cdap-ui/app/cdap/components/Tags/index.js [hunk 3]
```diff
@@ -15,33 +15,123 @@
...
+  fetchUserTags() {
+    let params = Object.assign({}, this.params, { scope: 'USER' });
+
+    MyMetadataApi.getTags(params)
+      .subscribe((res) => {
+        this.setState({
+          userTags: res,
+        });
+      }, (err) => {
+        console.log('Error', err);
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/Tags/index.js]: Same here too.

File: cdap-ui/app/cdap/components/Tags/index.js [hunk 4]
```diff
@@ -15,33 +15,123 @@
...
+      });
+  }
+  addTag() {
+    if (this.state.currentInputTag !== '') {
+      MyMetadataApi.addTags(this.params, [this.state.currentInputTag])
+      .subscribe(() => {
+        this.fetchUserTags();
+        this.toggleInputField();
+      }, (err) => {
+        console.log(err);
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/Tags/index.js]: Same here too. If users are not able to add a tag then we should show appropriate error message for failure.

File: cdap-ui/app/cdap/components/SpotlightSearch/SpotlightModal/SpotlightModalHeader.js [hunk 5]
```diff
@@ -43,10 +43,14 @@ require('./SpotlightModal.scss');
        <ModalHeader>
          <span className="float-xs-left">
            {
-
-             T.translate('features.SpotlightSearch.SpotlightModal.headerSearchResults', {
-               query: this.props.query
-             })
+             this.props.tag ?
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/SpotlightSearch/SpotlightModal/SpotlightModalHeader.js]: I don't think we need this check. Spotlight search is not being used for anything other than `shows list of entities matching a tags` as of now. The `query` message was for the spotlight search when used for searching entities.

**Review by ajainarayanan — COMMENTED** [review]
Couple of comments. Otherwise code looks good to me.

**Comment by tonybach** [comment]
@ajainarayanan Ok I've surfaced the error when there is one, but shouldn't the content of the error message come from the backend? Let me know if otherwise.

File: cdap-ui/app/cdap/components/Tags/Tag/index.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    this.setState({
+      searchModalOpen: !this.state.searchModalOpen
+    });
+  }
+
+  render() {
+    return (
+      <span>
+        {
+          this.props.scope === 'SYSTEM' ?
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Tags/Tag/index.js]: might be able to simplify this tertiary operation by using `classnames` to conditionally add the user or system-tags class, then simply show or hide the close button

**Comment by ajainarayanan** [comment]
@tonybach Right now, if the backend takes time to respond then adding tags is not giving the proper visual indication that something is being added. Or for that matter if we add it, it either appears right away or appears in the front of the list that it takes time to recognize that something has been added.

**Comment by tonybach** [comment]
Build: http://builds.cask.co/browse/CDAP-DRC5678

File: cdap-ui/app/cdap/components/SpotlightSearch/index.js [hunk 7]
```diff
@@ -35,7 +35,7 @@ const keyMap = {
   arrowDown: 40
 };
 
-const VIEW_RESULT_LIMIT = 5;
+const VIEW_RESULT_LIMIT = 0;
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/SpotlightSearch/index.js]: Is this change still necessary?

Comment (tonybach) [reply cdap-ui/app/cdap/components/SpotlightSearch/index.js]: Nice catch. I've reverted the change.

File: cdap-ui/app/cdap/components/Tags/index.js [hunk 8]
```diff
@@ -15,33 +15,232 @@
...
+  renderSystemTags() {
+    return (
+      <span>
+        {
+          this.state.systemTags.map(tag => {
+            return (
+              <Tag
+                value = {tag}
+                scope = 'SYSTEM'
+
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/Tags/index.js]: extra empty line

**Comment by elfenheart** [comment]
Should we limit the number of tags we show by default? and have a "Show More" perhaps?

**Comment by tonybach** [comment]
@elfenheart We can. We have that in the old UI right? @bdmogal What would the limit be?

**Comment by tonybach** [comment]
@elfenheart I talked with @bdmogal, and we agreed that it can be part of a general tags improvement JIRA to be done in 4.2, since Nitin had some other suggestions about it as well.

**Comment by elfenheart** [comment]
👍  LGTM

**Comment by tonybach** [comment]
Green build, merging.
